notte/pipe/main.py

```diff
@@ -46,3 +48,51 @@ def forward(
 
         actions = ActionFilteringPipe.forward(context, actions)
         return ActionSpace(_actions=actions)
+
+    def forward(
+        self,
+        context: Context,
+        previous_action_list: list[Action] | None = None,
+        n_trials: int = 2,
+        tresh_complete: float = 0.95,
+        gamma: int = 12000,  # num chars threshold. (token estimated.)
+    ) -> ActionSpace:
+        # triggers NotteNode[._chars,._path] calculation at build time.
+        _ = context.format(context.node)
+
+        # if the node size fits, we can proceed with the action listing.
+        if context.node._chars <= gamma:
+            return self._forward(context, previous_action_list, n_trials, tresh_complete)
+
+        # if the node is too big, we split it into subnodes.
+        subnodes = split(context.node, gamma)
+        chars = [subnode._chars for subnode in subnodes]
+        paths = [subnode._path for subnode in subnodes]
+        partitions = partition(chars, gamma)
+
+        trees: list[NotteNode] = []
+        for _partition in partitions:
+            if len(_partition) > 1:
+                _paths = [paths[i] for i in _partition]
+                partials = [partial_tree_by_path(context.node, path) for path in _paths]
+                merged = merge_trees(partials)
+                if merged is not None:
+                    trees.append(merged)
+            else:
+                path = paths[_partition[0]]
+                partial = partial_tree_by_path(context.node, path)
+                trees.append(partial)
+
+        # _forward for each tree, with incremental listing.
```

Comment:
I think that it's better to put lines `60-86` in a function `build_context_partitions`  to increase readability / lisibility

Reply:
Yes and it will also be helpful for you in data scraping 👍🏽